java/dagger/internal/codegen/BindingGraph.java

```diff
@@ -1120,6 +1122,24 @@ private boolean hasLocallyPresentOptionalBinding(ResolvedBindings resolvedBindin
               && !getLocalExplicitBindings(keyFactory.unwrapOptional(resolvedBindings.key()).get())
                   .isEmpty();
         }
+
+        /**
+         * Returns {@code true} if {@code resolvedBindings} contains a provision binding
+         * for which there is an explicit present binding in this component that was resolved
+         * previously.
+         */
+        boolean hasLocallyConflictingContributionBinding(ResolvedBindings resolvedBindings) {
+          return resolvedBindings
+                .contributionBindings()
+                .stream()
+                .anyMatch(contributionBinding -> contributionBinding.bindingKind().equals(PROVISION)
+                  && !getLocalExplicitBindings(resolvedBindings.key()).isEmpty()
+                  && parentResolver.isPresent()
```

Comment:
> Or am I missing something?

`dependsOnLocalBindings()` is also used in `getOwningComponent()`. 

The trouble is we’ve now changed dependsOnLocalBindings() to return `true` if any dependencies of a binding are defined locally. When foo is resolved in `Child`  the owning component is now `Child` when it should be `Parent`.

By checking the parent’s previously resolved bindings we know that this binding was not resolved in `Parent` and hence should not be owned by `Child` . 

 `GraphValidationTest.bindingUsedOnlyInSubcomponentDependsOnBindingOnlyInSubcomponent` fails without this check.

Thoughts?